### PR TITLE
Adding init_label_set method from prometheus ruby client

### PR DESCRIPTION
**Pull request by AntoineC44**
Hello,

Preview to get feedback, will add tests if accepted on principle.

Following https://github.com/fluent/fluent-plugin-prometheus/pull/180 here is a PR to add the `init_label_set` feature from prometheus ruby client (see https://github.com/prometheus/client_ruby/commit/0e0f17fb38f2e703f8adfaf16d75b26af32b5053)

Here is what it gives with this implem

fluent.conf
```xml
 <source>
    @type tcp
    format json
    bind "0.0.0.0"
    port 5170
    tag "test"
    <parse>
      @type json
    </parse>
  </source>
  <match test>
    @type prometheus
    <metric>
      name histogram
      type histogram
      desc histogram with current implem, not initialized until first event
      key test
      <labels>
        label a
      </labels>
    </metric>
    <metric>
      name histogram_initialized
      type histogram
      desc histogram initialized
      key test
      buckets 0, 1, 10, 1000
      <labels>
        iam initialized!
      </labels>
      <initlabels>
        iam initialized!
      </initlabels>
    </metric>
    <metric>
      name alert_counter_initalized
      type counter
      desc User knowing part of/all future values of reccordAccessor $.filename can initialize the corresponding counters
      <labels>
        filename $.filename
      </labels>
      <initlabels>
        filename foo.txt
      </initlabels>
      <initlabels>
        filename foo.tar.gz
      </initlabels>
    </metric>
    <metric>
      name simple_summary_initialized
      type summary
      desc Providing no initlabels in the block yields a simple initialized metric
      key test
      <initlabels>
      </initlabels>
    </metric>
  </match>
  <match>
    @type prometheus
    <metric>
      name counter
      type counter
      desc string labels must be identical between labels and initlabels blocks
      key test
      <labels>
        source fluentd
        filename $.filename
      </labels>
      <initlabels>
        source fluentd
        filename foo.tar.gz
      </initlabels>
      <initlabels>
        source fluentd
        filename foo.txt
      </initlabels>
    </metric>
  </match>
  <match>
    @type prometheus
    <labels>
      globallabel initvalue2
    </labels>
    <metric>
      name counter_with_global_label
      type counter
      desc Identically, global labels must be initialized in initlabels block
      <labels>
        locallabel $.key
      </labels>
      <initlabels>
        locallabel initvalue
        globallabel initvalue2
      </initlabels>
    </metric>
  </match>
  <source>
    @type prometheus
  </source>
```

result
```bash
$ curl http://localhost:24231/metrics
# TYPE histogram histogram
# HELP histogram histogram with current implem, not initialized until first event
# TYPE histogram_initialized histogram
# HELP histogram_initialized histogram initialized
histogram_initialized_bucket{iam="initialized!",le="0"} 0.0
histogram_initialized_bucket{iam="initialized!",le="1"} 0.0
histogram_initialized_bucket{iam="initialized!",le="10"} 0.0
histogram_initialized_bucket{iam="initialized!",le="1000.0"} 0.0
histogram_initialized_bucket{iam="initialized!",le="+Inf"} 0.0
histogram_initialized_sum{iam="initialized!"} 0.0
histogram_initialized_count{iam="initialized!"} 0.0
# TYPE alert_counter_initalized counter
# HELP alert_counter_initalized User knowing part of/all future values of reccordAccessor $.filename can initialize the corresponding counters
alert_counter_initalized{filename="foo.txt"} 0.0
alert_counter_initalized{filename="foo.tar.gz"} 0.0
# TYPE simple_summary_initialized summary
# HELP simple_summary_initialized Providing no initlabels in the block yields a simple initialized metric
simple_summary_initialized_sum 0.0
simple_summary_initialized_count 0.0
# TYPE counter counter
# HELP counter string labels must be identical between labels and initlabels blocks
counter{source="fluentd",filename="foo.tar.gz"} 0.0
counter{source="fluentd",filename="foo.txt"} 0.0
# TYPE counter_with_global_label counter
# HELP counter_with_global_label Identically, global labels must be initialized in initlabels block
counter_with_global_label{locallabel="initvalue",globallabel="initvalue2"} 0.0
```

Example of things that triggers configError with this implem

Having initlabels not matching a string value from label (metric would never be used)
```xml
  <match test>
    @type prometheus
    <metric>
      name counter
      type counter
      desc histogram with current implem, not initialized until first event
      <labels>
        label a
      </labels>
      <initlabels>
        label b ---> would trigger error because value a is expected on whole lifetime of metric
      </initlabels>
    </metric>
  </match>
```

Having initlabels not matching the signature of labels
```xml
  <match test>
    @type prometheus
    <metric>
      name counter
      type counter
      desc histogram with current implem, not initialized until first event
      <labels>
        somelabel a
        anotherlabel b
      </labels>
      <initlabels>
        somelabel a
        someinitlabel b ---> would trigger error because someinitlabel is not present in <labels> block, this has no meaning
      </initlabels>
    </metric>
  </match>
```

### CODE REVIEW

**Comment by AntoineC44** [comment]
Hello @repeatedly, would appreciate your feedback.
Adds a layer of complexity but fully retrocompatible and can be useful.
Cheers

**Comment by AntoineC44** [comment]
Hello @cosmo0920 @repeatedly,
Kindly requesting some feedback from you guys on this feature and implementation.
If ok for you I would add some dedicated tests,
Have a good day,
Antoine

**Comment by amaury-d** [comment]
Thanks @AntoineC44 for this pull request. I confirm this feature would be very useful too on our side.

**Comment by AntoineC44** [comment]
Hello @cosmo0920 @repeatedly @ashie,
Sorry to ping you again, are you able to provide some feedback?
If ok for you I would add some dedicated tests,
Have a good day,
Antoine

**Comment by AntoineC44** [comment]
Hello,
I wanted some feedback with the implem from a maintainer before writing tests and having the PR ready,
No feedback for now, I can retag @cosmo0920 @repeatedly @ashie,
Thanks,
Antoine

**Review by ashie — CHANGES_REQUESTED** [review]
Sorry for my too late response.
It looks good to me, thanks for your proposal!
Please address comments I added in-line.

**Comment by ashie** [comment]
In addition, could you add description about new config parameter to README.md?

File: lib/fluent/plugin/prometheus.rb [hunk 1]
```diff
@@ -165,6 +186,29 @@ def initialize(element, registry, labels)
...
+               if v.is_a?(String)
+                 if initlabel[k] != v
+                   raise ConfigError, "initlabel '#{k}' for metric '#{metric.name}' cannot have value '#{initlabel[k]}' " \
+                                      "different from provided label string value '#{v}'. " \
+                                      "Initialized metric would never be used."
+                 end
+               end
+             end
+             metric.init_label_set(initlabel)
+            }
```
[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:211]: Well spotted

File: lib/fluent/plugin/prometheus.rb [hunk 2]
```diff
@@ -56,6 +56,27 @@ def self.parse_labels_elements(conf)
         base_labels
       end
 
+      def self.parse_initlabels_elements(conf)
+        base_initlabels = []
+
+        conf.elements.select { |e| e.name == 'initlabels' }.each { |block|
+          initlabels = {}
+          block.each do |key, value|
+            block.has_key?(key)
```
[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:65]: Indeed

File: lib/fluent/plugin/prometheus.rb [hunk 3]
```diff
@@ -165,6 +186,29 @@ def initialize(element, registry, labels)
...
+             end
+             
+             base_labels.each do |k, v|
+               if v.is_a?(String)
+                 if initlabel[k] != v
+                   raise ConfigError, "initlabel '#{k}' for metric '#{metric.name}' cannot have value '#{initlabel[k]}' " \
+                                      "different from provided label string value '#{v}'. " \
+                                      "Initialized metric would never be used."
+                 end
+               end
```
[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:208]: Had the same thought for a boolean but ended up doing this because of the record accessors so as to keep only one way to configure initlabels, WDYT?

[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:208]: We could make a mix, initialization boolean for all string labels and have a different config for record accessors labels and with a config error should a reccord accessor with initialization set to true does not specify init labels blocs?

```
    <metric>
      name counter
      type counter
      desc string
      key test
      initialized true
      <labels>
        source fluentd
      </labels>
    </metric>
    <metric>
      name foo
      type counter
      desc string
      initialized true
      <labels>
        regularlabel initvalue
        recordaccessor $.key
      </labels>
      <recordaccessorinitlabels>
        recordaccessor initvalue1
      </recordaccessorinitlabels>
      <recordaccessorinitlabels>
        recordaccessor initvalue2
      </recordaccessorinitlabels>
    </metric>
```

File: lib/fluent/plugin/prometheus.rb [hunk 4]
```diff
@@ -165,6 +186,29 @@ def initialize(element, registry, labels)
...
           @base_labels = Fluent::Plugin::Prometheus.parse_labels_elements(element)
           @base_labels = labels.merge(@base_labels)
+
+          @base_initlabels = Fluent::Plugin::Prometheus.parse_initlabels_elements(element)
+        end
+
+        def self.init_label_set(metric, base_initlabels, base_labels)
+           base_initlabels.each { |initlabel|
+             if initlabel.keys.sort != base_labels.keys.sort
+               raise ConfigError, "initlabels for metric #{metric.name} must have the same signature than labels " \
```
[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:196]: Good point, no errors showed up during tests so we don't have a test for that?

Don't know how to test easily but this would suffice?

```
      def self.parse_initlabels_elements(conf)
        base_initlabels = []

        conf.elements.select { |e| e.name == 'initlabels' }.each { |block|
          initlabels = {}
          block.each do |key, value|
            # A RecordAccessor has no meaning for initializing a metric's labels
            if value.start_with?('$.') || value.start_with?('$[')
              raise ConfigError, "RecordAccessor cannot be used in an initlabel"
            else 
              initlabels[key.to_sym] = value
            end
          end

          if @multi_worker
            base_labels[:worker_id] = fluentd_worker_id.to_s
          end
          
          base_initlabels << initlabels
        }

        base_initlabels
      end
```

[the remark this answers is not in the document]
Comment (ashie) [reply lib/fluent/plugin/prometheus.rb:196]: > Don't know how to test easily but this would suffice?

Yeah, I think so.

File: lib/fluent/plugin/prometheus.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashie) [reply lib/fluent/plugin/prometheus.rb:208]: > We could make a mix, initialization boolean for all string labels and have a different config for record accessors labels and with a config error should a reccord accessor with initialization set to true does not specify init labels blocs?

Thanks, it looks better to me than before.
`recordaccessorinitlabels` should be splitted with `_`.

[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb]: Forgot about the existence of Place Holders, new implem handles both Place Holders and Reccord Accessors, so I ended up re-using `<initlabels>`

**Comment by AntoineC44** [comment]
Didi the changes, it indeed ends-up being less configuration but more code, especially to handle corner cases around placeholders and reccord accessors.

**Comment by AntoineC44** [comment]
Fixed typo in README and not broad enough test for the config

File: lib/fluent/plugin/prometheus.rb [hunk 5]
```diff
@@ -53,7 +53,69 @@ def self.parse_labels_elements(conf)
           end
         end
 
-        base_labels
+        base_labels
```
[the remark this answers is not in the document]
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb]: Updated

**Comment by AntoineC44** [comment]
@ashie Hello any news?

File: lib/fluent/plugin/prometheus.rb [hunk 6]
```diff
@@ -97,6 +159,10 @@ def stringify_keys(hash_to_stringify)
         end.to_h
       end
 
+      def self.set_worker_id(worker_id)
+        @worker_id = worker_id
+      end
+
```
Comment (ashie) [inline lib/fluent/plugin/prometheus.rb:165]: Is this method needed?
`Fluent::Plugin::Prometheus` is intend to be included from plugins so that it can use `fluentd_woker_id` directly.
`Fluent::Plugin::PrometheusLabelParser` is in also similar context and it uses `fluentd_woker_id` directly in fact.

I think the reason you added it is that you defined `Fluent::Plugin::Prometheus.parse_initlabels_elements` as a class method, but I think it's not needed to define as class method. Defining it as an instance method will simplify the code.

File: lib/fluent/plugin/prometheus.rb [hunk 7]
```diff
@@ -56,6 +56,68 @@ def self.parse_labels_elements(conf)
         base_labels
       end
 
+      def self.parse_initlabels_elements(conf, base_labels)
```
Comment (ashie) [inline lib/fluent/plugin/prometheus.rb:59]: How about removing the prefix `self.` to make it as instance method?
So that you can use `fluentd_worker_id` directly instead of setting `@worker_id` via `Fluent::Plugin::Prometheus.set_worker_id`.

**Review by ashie — CHANGES_REQUESTED** [review]

File: lib/fluent/plugin/prometheus.rb (hunk 7 above)
Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:59]: Hello,
Seems to me the new method `parse_initlabels_elements` has to be called like `parse_labels_elements` in `Metric` class in its `initialize` method (around line 230), hence it is a class method we need not an instance method
And yes from the moment it is a class method we need this (dirty) trick to pass down the `fluentd_worker_id` :/

Do you see a way of writing that without droping support for `fluentd_worker_id` or having to reorganize the code?

Comment (ashie) [reply lib/fluent/plugin/prometheus.rb:59]: Ah, I see. Sorry I overlooked it.

> Do you see a way of writing that without droping support for fluentd_worker_id or having to reorganize the code?

I think passing it by argument is proper way, but such change seems overkill in this case.

The worker ID is determined per process, can be retrieved from `ENV['SERVERENGINE_WORKER_ID']`.

https://github.com/fluent/fluentd/blob/c0f1db966a2809a13de217b0ec0adc9a29decc79/lib/fluent/plugin/base.rb#L51

```ruby
      def fluentd_worker_id
        return @_fluentd_worker_id if @_fluentd_worker_id
        @_fluentd_worker_id = (ENV['SERVERENGINE_WORKER_ID'] || 0).to_i
        @_fluentd_worker_id
      end
```

I think using `ENV['SERVERENGINE_WORKER_ID']` is enough for this situation.

Or, using [`Fluent::VariableStore`](https://github.com/fluent/fluentd/blob/c0f1db966a2809a13de217b0ec0adc9a29decc79/lib/fluent/variable_store.rb
) instead of class variable or class-instance variable is recommended in fluentd plugins.

Comment (AntoineC44) [reply lib/fluent/plugin/prometheus.rb:59]: I liked the `ENV['SERVERENGINE_WORKER_ID']` idea, I replaced by this
Let me know if you see anything else to change

**Review by ashie — APPROVED** [review]
LGTM

I'm sorry for long long pending.
And thank you for your effort!
